rpcclient/infrastructure.go

```diff
@@ -1733,48 +1733,35 @@ func (c *Client) Send() error {
 	return nil
 }
 
+// Copied from newer Go version
+func cutPrefix(s, prefix string) (after string, found bool) {
+	if !strings.HasPrefix(s, prefix) {
+		return s, false
+	}
+	return s[len(prefix):], true
+}
+
 // ParseAddressString converts an address in string format to a net.Addr that is
 // compatible with btcd. UDP is not supported because btcd needs reliable
 // connections. We accept a custom function to resolve any TCP addresses so
 // that caller is able control exactly how resolution is performed.
 func ParseAddressString(strAddress string) (net.Addr, error) {
-	var parsedNetwork, parsedAddr string
-
-	// Addresses can either be in network://address:port format,
-	// network:address:port, address:port, or just port. We want to support
-	// all possible types.
-	if strings.Contains(strAddress, "://") {
-		parts := strings.Split(strAddress, "://")
-		parsedNetwork, parsedAddr = parts[0], parts[1]
-	} else if strings.Contains(strAddress, ":") {
-		parts := strings.Split(strAddress, ":")
-		parsedNetwork = parts[0]
-		parsedAddr = strings.Join(parts[1:], ":")
-	} else {
-		parsedAddr = strAddress
+	// Addresses can either be in unix://address, unixpacket://address URL format
+	// Or just address:port host format for tcp.
+	if after, ok := cutPrefix(strAddress, "unix://"); ok {
+		return net.ResolveUnixAddr("unix", after)
 	}
-
-	// Only TCP and Unix socket addresses are valid. We can't use IP or
-	// UDP only connections for anything we do in lnd.
-	switch parsedNetwork {
-	case "unix", "unixpacket":
-		return net.ResolveUnixAddr(parsedNetwork, parsedAddr)
-
-	case "tcp", "tcp4", "tcp6":
-		return net.ResolveTCPAddr(parsedNetwork, verifyPort(parsedAddr))
-
-	case "ip", "ip4", "ip6", "udp", "udp4", "udp6", "unixgram":
-		return nil, fmt.Errorf("only TCP or unix socket "+
-			"addresses are supported: %s", parsedAddr)
-
-	default:
-		// We'll now possibly use the local host short circuit
-		// or parse out an all interfaces listen.
-		addrWithPort := verifyPort(strAddress)
-
-		// Otherwise, we'll attempt to resolve the host.
-		return net.ResolveTCPAddr("tcp", addrWithPort)
+	if after, ok := cutPrefix(strAddress, "unixpacket://"); ok {
+		return net.ResolveUnixAddr("unixpacket", after)
+	}
+	if strings.Contains(strAddress, "://") { // Not supporting :// anywhere in the host or path
+		return nil, fmt.Errorf("unsupported protocol in address: %s", strAddress)
+	}
+	u, err := url.Parse("dummy://" + strAddress)
+	if err != nil {
+		return nil, err
 	}
+	return net.ResolveTCPAddr("tcp", verifyPort(u.Host))
```

Comment:
I don't actually understand how this fixes your issue. If you say you're using `somehost/path`, then `/path` would be removed here. Which makes me think that using `.../path` is wrong anyway?
What's the use case here? Is the RPC server behind a reverse proxy that routes traffic according to paths?

Reply:
When we previously had path here - net.ResolveTCPAddr failed, now it succeeds, since it only needs the host.

The path is then later actually used in:

```
var url string
	switch parsedAddr.Network() {
	case "unix", "unixpacket":
		// Using a placeholder URL because a non-empty URL is required.
		// The Unix domain socket is specified in the DialContext.
		url = protocol + "://unix"
	default:
		url = protocol + "://" + c.config.Host
	}
...
	for i := 0; i < tries; i++ {
...
		httpReq, err = http.NewRequest("POST", url, bodyReader)
...
```

Reply:
We could have restricted path anyway in `config.Host`, and perhaps requested that it would be used in `config.Endpoint` which is actually used as path in websocket flows - but that would be breaking behavior.

Plus now `Host` starts to contain prefixes such as `unix://` where previously (and still now) `http://` was not allowed.

The cleanest way to do it obviously IMO is to have:
```
type ConnConfig struct {

	Scheme string

	Host string

	Endpoint string

}
`

And use it:

```
Scheme: "unix"
Host: "host-no-path:123"
Endpoint: "/path/to/call"
```

And throw all the URL parsing logic away.

Reply:
Yeah, I guess using the endpoint would've been the cleanest option. But that would be breaking to users that previously specified the path in the `Host`.

Reply:
However, we now also added the usage of `unix://` in the `Host` part - which will also break in the future if we introduce `Scheme`.

Might consider breaking the existing *new* unix socket users in favor of not breaking them later on once this gets more popular by adding a `Scheme`.

Regarding breaking existing `Host` users - that's why I wrote my PR this way, however it's tempting to breaking them as well since the documentation of the module clearly says host only - but I don't mind

Reply:
Let's not do a breaking change now but instead fix the issue introduced by #2168. See my diff/proposal below.

Reply:
Ok, so I added your cleanups and added some tests (hopefully they do not need extra cleanups).